### ISSUE TITLE
Printed expression is not invariant under evaluation

### ISSUE THREAD

**Issue by sylee957** [sympy/sympy#20616]
```r
from sympy import *

x = Symbol('x')
l = Symbol('lambda')

expr = -l/(x - 1)
after = expr.subs(l, x - sqrt(2)*(x - 1) + 1)
eval(str(after), {'x': x, 'sqrt': sqrt}) == after
```
The original expression gets printed with `-(x - sqrt(2)*(x - 1) + 1)/(x - 1)`, but if you doubly execute the string, the minus sign gets into the numerator and it becomes `(-x + sqrt(2)*(x - 1) - 1)/(x - 1)`

**Comment by harshavardhan-99** [comment]
I would like to work on this issue.

**Comment by oscarbenjamin** [comment]
The issue is the 2-arg Mul:
```
In [1]: Mul(2, x+y)
Out[1]: 2⋅x + 2⋅y

In [2]: Mul(2, x+y, z)
Out[2]: 2⋅z⋅(x + y)
```
Specifically a Mul with 2-args where one arg is a Number and the other is an Add will distribute.

When you do something like this:
```
In [3]: 2*(x + y)*z
Out[3]: z⋅(2⋅x + 2⋅y)
```
It is evaluated as
```
In [4]: Mul(Mul(2, x+y), z)
Out[4]: z⋅(2⋅x + 2⋅y)
```
We should change the repr printer to add extra brackets to prevent this:
```
In [5]: 2*((x + y)*z)
Out[5]: 2⋅z⋅(x + y)
```
The repr printer should check for a Mul with more than 2 args and if the first arg is a `Number` and the second (in display order) is an `Add` then everything after the first arg should be parenthesised. That would fix the problem for any Mul with more than 2 args.

In the case where the Mul actually has 2 args there is no way to fix this without using `Mul(a, b, evaluate=False)` which we probably do not want in the standard repr display. There could be a printing option though that is for "code printing" that does show this with `evaluate=False` and also prints rationals as e.g. `S(1)/2`.

**Comment by lastcodestanding** [comment]
Is anyone working on this at the moment? 
If not, I'd like to resolve it.

**Comment by lastcodestanding** [comment]
I think this has been done intentionally. the `flatten()` function says that it distributes the 2-arg `Mul()`, and simultaneously processes 3 or more args.

**Comment by oscarbenjamin** [comment]
Yes, this was intentional although now I think it is agreed that distributing the 2-arg mul is a mistake (it's not easy to change though).

We can still fix the problem that the printed output does not evaluate to the same result in many cases by adding brackets.
